From c89795545d24bc11ee9b9e7730b72d5909ea3194 Mon Sep 17 00:00:00 2001
Message-Id: <c89795545d24bc11ee9b9e7730b72d5909ea3194.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: Shashank Sharma <shashank.sharma@intel.com>
Date: Wed, 28 Jan 2015 15:48:55 +0530
Subject: [PATCH 31/95] MUST_REBASE [VPG]: adf/intel/hdmi: HDMI hotplug
 handler

This patch adds HDMI hotplug handler code for the encoder side and
also adds few registers to check HDMI interrupt and live status.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

Issue: GMINL-5482
Change-Id: Ifb478c268274000c17dad183ceacc559d3626a42
Signed-off-by: Shashank Sharma <shashank.sharma@intel.com>
---
 drivers/video/adf/intel/core/common/hdmi/Makefile  |    1 +
 .../adf/intel/core/common/hdmi/gen_hdmi_pipe.c     |   13 +++++++++++++
 .../video/adf/intel/include/core/vlv/chv_dc_regs.h |    6 ++++--
 3 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/drivers/video/adf/intel/core/common/hdmi/Makefile b/drivers/video/adf/intel/core/common/hdmi/Makefile
index eb1edba..5e9989e 100644
--- a/drivers/video/adf/intel/core/common/hdmi/Makefile
+++ b/drivers/video/adf/intel/core/common/hdmi/Makefile
@@ -8,4 +8,5 @@ ifdef CONFIG_ADF_INTEL_VLV
 	obj-y := gen_hdmi_pipe.o
 	obj-y += gen_hdmi_pipe_ops.o
 	obj-y += gen_hdmi_edid.o
+	obj-y += gen_hdmi_hotplug.o
 endif
diff --git a/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c b/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c
index 6ce2562..39d4a3d 100644
--- a/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c
+++ b/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c
@@ -69,6 +69,7 @@ int hdmi_context_init(struct hdmi_context *context)
 {
 	struct hdmi_config *hdmi_config;
 	struct hdmi_pipe *hdmi_pipe;
+	int ret;
 
 	if (!context) {
 		/* How ?? */
@@ -89,6 +90,13 @@ int hdmi_context_init(struct hdmi_context *context)
 		return -EINVAL;
 	}
 
+	/* Probe the current HDMI status */
+	ret = intel_adf_hdmi_probe(hdmi_pipe, true);
+	if (ret) {
+		pr_err("ADF: HDMI: %s Probing HDMI failed\n", __func__);
+		return ret;
+	}
+
 #ifdef INTEL_ADF_HDMI_SELF_MODESET_AT_BOOT
 	/* If hdmi connected, do modeset on pref mode */
 	if (atomic_read(&context->connected)) {
@@ -185,6 +193,11 @@ int hdmi_pipe_init(struct hdmi_pipe *pipe,
 	/* Fixme: Hardcoding bpp */
 	pipeline->params.hdmi.bpp = 24;
 
+	/* Load HDMI interface ops */
+	pipe->ops.set_event = intel_adf_hdmi_set_events;
+	pipe->ops.get_events = intel_adf_hdmi_get_events;
+	pipe->ops.handle_events = intel_adf_hdmi_handle_events;
+	pipe->ops.get_hw_state = intel_adf_hdmi_get_hw_events;
 	pipe->dpms_state = DRM_MODE_DPMS_OFF;
 
 	/* Init the PIPE */
diff --git a/drivers/video/adf/intel/include/core/vlv/chv_dc_regs.h b/drivers/video/adf/intel/include/core/vlv/chv_dc_regs.h
index 3123015..6cb8848 100644
--- a/drivers/video/adf/intel/include/core/vlv/chv_dc_regs.h
+++ b/drivers/video/adf/intel/include/core/vlv/chv_dc_regs.h
@@ -24,8 +24,10 @@
 #define CHV_HPD_STAT	(CHV_DISPLAY_BASE + 0x61114)
 #define CHV_HPD_LIVE_STATUS_MASK	(0x7 << 27)
 #define CHV_HPD_LIVE_STATUS_B	(0x1 << 29)
-#define CHV_HPD_LIVE_STATUS_C	(0x7 << 28)
-#define CHV_HPD_LIVE_STATUS_D	(0x7 << 27)
+#define CHV_HPD_LIVE_STATUS_C	(0x1 << 28)
+#define CHV_HPD_LIVE_STATUS_D	(0x1 << 27)
+#define CHV_HPD_LIVE_STATUS(port) (CHV_HPD_LIVE_STATUS_B \
+				>> (port - PORT_B))
 
 /* CHV SDVO/HDMI bits: */
 #define   SDVO_PIPE_SEL_CHV(pipe)		((pipe) << 24)
-- 
1.7.9.5

